<a href="https://colab.research.google.com/github/Yashraj2050/Data_Science_Lab_SE_A_34/blob/main/prac3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Use NumPy to perform matrix operations including calculating the determinant of a matrix, finding its inverse and verifying it, and solving a system of linear equations.

## Introduce NumPy for Matrix Manipulations

### Subtask:
Demonstrate the creation of matrices using NumPy and perform basic matrix operations to illustrate its capabilities for matrix manipulations.


**Reasoning**:
First, I'll import the NumPy library to enable matrix manipulations. This is a foundational step for any operations involving arrays or matrices in Python.



In [1]:
import numpy as np
print("NumPy imported successfully.")

NumPy imported successfully.


**Reasoning**:
Now that NumPy is imported, I will create two sample matrices, perform matrix addition and multiplication, and print the results to demonstrate basic matrix operations as per the instructions.



In [2]:
matrix_a = np.array([[1, 2], [3, 4]])
matrix_b = np.array([[5, 6], [7, 8]])

print("Matrix A:\n", matrix_a)
print("Matrix B:\n", matrix_b)

# Matrix Addition
matrix_sum = matrix_a + matrix_b
print("\nMatrix A + Matrix B:\n", matrix_sum)

# Matrix Multiplication
matrix_product = np.dot(matrix_a, matrix_b)
# Alternatively, using the @ operator:
# matrix_product = matrix_a @ matrix_b
print("\nMatrix A * Matrix B (using np.dot):\n", matrix_product)

Matrix A:
 [[1 2]
 [3 4]]
Matrix B:
 [[5 6]
 [7 8]]

Matrix A + Matrix B:
 [[ 6  8]
 [10 12]]

Matrix A * Matrix B (using np.dot):
 [[19 22]
 [43 50]]


## Calculate Matrix Determinant

### Subtask:
Define a sample square matrix and calculate its determinant using NumPy's linear algebra functions. Display the original matrix and its determinant.


**Reasoning**:
To fulfill the subtask, I will define a sample square matrix using `np.array`, calculate its determinant using `np.linalg.det`, and then print both the matrix and its determinant.



In [3]:
det_matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

det_value = np.linalg.det(det_matrix)

print("Original Matrix:", det_matrix)
print("Determinant of the matrix:", det_value)

Original Matrix: [[1 2 3]
 [4 5 6]
 [7 8 9]]
Determinant of the matrix: 0.0


## Find Matrix Inverse

### Subtask:
Define a sample invertible square matrix, calculate its inverse using NumPy, and verify the inverse by multiplying it with the original matrix to check for an identity matrix. Display the original matrix, its inverse, and the verification result.


#### Reasoning:
To fulfill the subtask, I will define an invertible matrix, calculate its inverse using `np.linalg.inv()`, verify the inverse by multiplying the original matrix by its inverse, and then print the original matrix, its inverse, and the verification result. This will demonstrate how to find and verify matrix inverses using NumPy.

**Reasoning**:
Following the instructions, I will define an invertible matrix, calculate its inverse, and then verify the inverse by multiplying it with the original matrix. Finally, I will print all the results as requested.



In [4]:
inv_matrix = np.array([[4, 7], [2, 6]])

inverse_matrix = np.linalg.inv(inv_matrix)

verification = np.dot(inv_matrix, inverse_matrix)

print("Original Matrix:\n", inv_matrix)
print("\nInverse Matrix:\n", inverse_matrix)
print("\nVerification (Original * Inverse):\n", verification)

Original Matrix:
 [[4 7]
 [2 6]]

Inverse Matrix:
 [[ 0.6 -0.7]
 [-0.2  0.4]]

Verification (Original * Inverse):
 [[ 1.00000000e+00 -1.11022302e-16]
 [ 1.11022302e-16  1.00000000e+00]]


## Solve System of Linear Equations

### Subtask:
Set up a system of linear equations in matrix form (Ax = b). Use NumPy's linear algebra functions to solve for x, the vector of unknowns. Display the coefficient matrix (A), the constant vector (b), and the solution vector (x).


**Reasoning**:
As per the subtask, I will define the coefficient matrix `A` and the constant vector `b`. Then, I will use `np.linalg.solve()` to find the solution vector `x` for the system Ax = b. Finally, I will print `A`, `b`, and `x` to display the problem setup and its solution.

**Reasoning**:
Following the reasoning, I will now define the coefficient matrix A and the constant vector b, then solve for x using `np.linalg.solve()` and print all the results as instructed.



In [5]:
A = np.array([[2, 1, -1], [1, 3, 2], [3, -2, 1]])
b = np.array([8, 1, 3])

x = np.linalg.solve(A, b)

print("Coefficient Matrix A:\n", A)
print("\nConstant Vector b:\n", b)
print("\nSolution Vector x:\n", x)

Coefficient Matrix A:
 [[ 2  1 -1]
 [ 1  3  2]
 [ 3 -2  1]]

Constant Vector b:
 [8 1 3]

Solution Vector x:
 [ 2.4  1.  -2.2]


## Final Task

### Subtask:
Provide a summary of the matrix operations performed and the results obtained for determinant, inverse, and solving linear equations using NumPy.


## Summary:

### Data Analysis Key Findings

*   **Matrix Initialization and Basic Operations:** NumPy was successfully used to create 2x2 matrices `A` (`[[1, 2], [3, 4]]`) and `B` (`[[5, 6], [7, 8]]`). Their sum was computed as `[[ 6, 8], [10, 12]]`, and their product (using `np.dot`) was `[[19, 22], [43, 50]]`.
*   **Determinant Calculation:** For the 3x3 matrix `[[1, 2, 3], [4, 5, 6], [7, 8, 9]]`, the determinant was calculated to be `0.0` using `np.linalg.det()`.
*   **Matrix Inverse and Verification:** The inverse of the 2x2 matrix `[[4, 7], [2, 6]]` was found to be `[[ 0.6, -0.7], [-0.2, 0.4]]` using `np.linalg.inv()`. Verification by multiplying the original matrix by its inverse yielded a result very close to an identity matrix, specifically `[[ 1.0e+00, -1.1e-16], [ 1.1e-16, 1.0e+00]]`, confirming the inverse calculation within floating-point precision.
*   **Solving Linear Equations:** A system of linear equations represented by the coefficient matrix `A` (`[[2, 1, -1], [1, 3, 2], [3, -2, 1]]`) and the constant vector `b` (`[8, 1, 3]`) was solved using `np.linalg.solve()`. The solution vector `x` was determined to be `[ 2.4, 1.0, -2.2]`.

### Insights or Next Steps

*   NumPy's `np.linalg` module provides efficient and accurate tools for common matrix operations, such as calculating determinants, inverses, and solving linear systems, which are fundamental in various scientific and engineering applications.
*   The determinant of `0.0` for one of the matrices indicates that it is singular (not invertible), which is a crucial concept in linear algebra and suggests that a system of equations with such a coefficient matrix would either have no unique solution or infinitely many solutions.
